### PR TITLE
10212: document send step

### PR DESCRIPTION
**Pull request by mkalish**
This PR adds documentation for the send step.

Test Steps:
n/a

## Changes
- Fills in the `send.md` page

### Testing
- [ ] Tested locally?
- [ ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #10212

### CODE REVIEW

File: prime-router/docs/universal-pipeline/send.md [hunk 1]
```diff
@@ -1,5 +1,126 @@
 # Send
 
-Retry mechanisms (in case of send failures)
+This is the final step in the pipeline and is responsible for actually delivering the batched report to a receivers
+preferred location
```
Comment (arnejduranovic) [inline prime-router/docs/universal-pipeline/send.md:4]:
```suggestion
configured location
```

File: prime-router/docs/universal-pipeline/send.md [hunk 2]
```diff
@@ -1,5 +1,126 @@
...
+  port: 22
+  filePath: ./upload
+  credentialName: DEFAULT-SFTP
+```
+
+## How it works
+
+This step works by listening for messages to get added to the send queue and then looking up the transport that is configured
+for the receiver.  Each transport has its own internal logic for creating the report and sending it out, but they mostly all follow
+the similar pattern of looking up the credentials for that transport/receiver, create an external file name and send the contents of the report.
```
Comment (arnejduranovic) [inline prime-router/docs/universal-pipeline/send.md:61]:
```suggestion
the similar pattern of looking up the credentials for that transport/receiver, creating an external file name and sending the contents of the report.
```

File: prime-router/docs/universal-pipeline/send.md [hunk 3]
```diff
@@ -1,5 +1,126 @@
...
+
+**Some important notes on the transports:**
+- The AS2 transport is unused
+- The SOAP transport is hardcoded to only work with the state of Pennsylvania
+- The Email transport is incomplete and does not work
+- The Rest Transport uses the KTOR library under the hood; if it encounters a `ClientRequestException` (any 4XX status), this transport wil short circuit and
```
Comment (arnejduranovic) [inline prime-router/docs/universal-pipeline/send.md:67]:
```suggestion
- The Rest Transport uses the KTOR library under the hood; if it encounters a `ClientRequestException` (any 4XX status), this transport will short circuit and
```

File: prime-router/docs/universal-pipeline/send.md [hunk 4]
```diff
@@ -1,5 +1,126 @@
...
+no retries will be attempted
+- The BlobStore transport only supports azure as the backing blob store and should only be used to move blobs internally in the CDC
+
+### Retries
+
+All the transports are configured to catch any thrown exception and return a retry token of all items.  When a retry token is returned, the SendFunction
+looks to see if it should re-queue the event for another attempt at getting sent.  The SendFunction will attempt to send the report 5 more times 
+using an exponential back off with a small, random amount of time added or subtracted; this is done in order to prevent a huge slew of send events all being 
+triggered at the same time in the event of a large outage.  See examples for more details.
```
Comment (arnejduranovic) [inline prime-router/docs/universal-pipeline/send.md:76]: `[examples](#examples)` I think will work here

File: prime-router/docs/universal-pipeline/send.md [hunk 5]
```diff
@@ -1,5 +1,126 @@
...
+parses the message and finds the relevant credentials in the vault. The function generates an external name and attempts to upload the file
+to the SFTP server, but it is currently down.  The SendFunction queues a new event with a visibility timeout of 8 minutes 
+(10 minutes since it is the first retry minus 2 minutes of random "dithering" time).  That second event is then read from the queue and the report is delivered.
+
+### A report fails to send five times and is never sent
+
+A receiver is configured to receive reports in a SFTP server and a new message is added to the send queue for that receiver.  The send queue
+parses the message and finds the relevant credentials in the vault. The function generates an external name and attempts to upload the file
+to the SFTP server, but gets a response that the credentials are invalid.  The function will then queue events with exponentially larger
+visibility timeouts until all five retries have occurred.  The function will then mark that the report failed to be sent.
```
[the remark this answers is not in the document]
Comment (arnejduranovic) [reply prime-router/docs/universal-pipeline/send.md:115]: Never mind this seems more appropriate in the troubleshooting doc or elsewhere, but keeping here to document my thoughts.

**Review by arnejduranovic — CHANGES_REQUESTED** [review]

**Comment by arnejduranovic** [comment]
Also, I think we should make a document or section somewhere that explains step-by-step how to set up the various transport types for a receiver and link it in this document at the end of the types section? With screenshots and everything - Then make a ticket for it and assign engagement? @mkalish thoughts?

**Comment by mkalish** [comment]
@arnejduranovic I dropped a link the receivers.md doc which will cover setting up a new transport

**Review by arnejduranovic — APPROVED** [review]

**Pull request merged**